File: backend-php/src/log_formatter.php
<?php

declare(strict_types=1);

function format_timestamp(mixed $value): string
{
    if ($value === null || $value === '') {
        return '';
    }
    $jst = new DateTimeZone(JST_TIMEZONE);
    if (is_int($value) || is_float($value)) {
        $date = DateTimeImmutable::createFromFormat('U.u', sprintf('%.6F', ((float) $value) / 1000), new DateTimeZone('UTC'));
        return $date ? $date->setTimezone($jst)->format('Y/m/d H:i:s') . ' JST' : (string) $value;
    }
    if (is_string($value)) {
        $trimmed = trim($value);
        $normalized = str_replace([' UTC', ' '], ['Z', 'T'], $trimmed);
        try {
            $date = new DateTimeImmutable(str_replace('Z', '+00:00', $normalized));
            return preg_match('/(?:Z|[+-]\d{2}:?\d{2})$/', $trimmed)
                ? $date->setTimezone($jst)->format('Y/m/d H:i:s') . ' JST'
                : $date->format('Y/m/d H:i:s') . ' JST';
        } catch (Exception) {
            return $trimmed;
        }
    }
    return (string) $value;
}

function detect_log_type(string $indexName): string
{
    if (str_contains($indexName, 'authlog')) {
        return 'authlog';
    }
    if (str_contains($indexName, 'syslog')) {
        return 'syslog';
    }
    return 'unknown';
}

function log_matches_exact_filters(array $log, array $filters): bool
{
    if ($filters['host'] !== '' && regex_pattern($filters['host']) === null && ($log['host'] ?? '') !== $filters['host']) {
        return false;
    }
    if ($filters['program'] !== '' && regex_pattern($filters['program']) === null && ($log['program'] ?? '') !== $filters['program']) {
        return false;
    }
    return true;
}

File: backend-java/src/main/java/com/example/flaskelastic/LogFormatter.java
package com.example.flaskelastic;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

final class LogFormatter {
    private static final ZoneId JST = ZoneId.of("Asia/Tokyo");
    private static final DateTimeFormatter DISPLAY_TIME =
            DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss 'JST'", Locale.ROOT);

    private LogFormatter() {}

    static String formatTimestamp(Object value) {
        if (value == null) return "";
        if (value instanceof Number number) {
            return Instant.ofEpochMilli(number.longValue()).atZone(JST).format(DISPLAY_TIME);
        }
        if (value instanceof String string) return formatTimestampString(string);
        return String.valueOf(value);
    }

    static String detectLogType(String indexName) {
        if (indexName != null && indexName.contains("authlog")) return "authlog";
        if (indexName != null && indexName.contains("syslog")) return "syslog";
        return "unknown";
    }

    static boolean matchesExactFilters(LogRecord log, Filters filters) {
        if (!filters.host().isBlank() && QueryBuilder.regexPattern(filters.host()) == null
                && !Objects.equals(log.host(), filters.host())) return false;
        return filters.program().isBlank() || QueryBuilder.regexPattern(filters.program()) != null
                || Objects.equals(log.program(), filters.program());
    }

    private static String formatTimestampString(String value) {
        String trimmed = value.trim();
        if (trimmed.isEmpty()) return "";
        for (String candidate : List.of(
                trimmed, trimmed.replace(" UTC", "Z"), trimmed.replace(" ", "T"),
                trimmed.replace(" UTC", "Z").replace(" ", "T")
        )) {
            try {
                return OffsetDateTime.parse(candidate).atZoneSameInstant(JST).format(DISPLAY_TIME);
            } catch (DateTimeParseException ignored) {}
            try {
                return LocalDateTime.parse(candidate).format(DISPLAY_TIME);
            } catch (DateTimeParseException ignored) {}
        }
        return value;
    }
}
